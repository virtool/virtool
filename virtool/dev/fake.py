"""
Code for creating a fake database and application data directory on jobs API startup.

TODO: Add more fake data and files
TODO: Port over to the normal API

"""

from logging import getLogger
from shutil import rmtree
from tempfile import mkdtemp

from sqlalchemy.ext.asyncio import AsyncSession

import virtool.analyses.db
import virtool.analyses.files
import virtool.users.db
import virtool.utils
import virtool.jobs.db
import virtool.users.db
from virtool.jobs.utils import JobRights
from virtool.types import App
from virtool.uploads.models import Upload
from virtool.utils import ensure_data_dir, random_alphanumeric

logger = getLogger(__name__)

USER_ID = "bob"


async def populate(app: App):
    await create_fake_user(app)
    await create_fake_subtractions(app)
    await create_fake_analysis(app)
    await create_fake_jobs(app)


async def remove_fake_data_path(app: App):
    """
    Remove the temporary fake application data directory created when the jobs API is run with the
    ``--fake`` option.

    :param app: the application object

    """
    data_path = app["config"].get("data_path")

    # Be extra sure this is a fake application directory we should remove.
    if data_path and app["config"]["fake"] and "virtool_fake_" in data_path:
        rmtree(data_path)
        logger.debug(f"Removed fake data directory: {data_path}")


async def drop_fake_mongo(app: App):
    """
    Drop a fake Mongo database if the instance was run with the ``--fake`` option.

    :param app: the application object

    """
    db_name = app["config"]["db_name"]

    # Be extra sure this is a fake database we should drop.
    if "fake-" in db_name and app["config"]["fake"]:
        await app["db"].motor_client.client.drop_database(db_name)
        logger.debug(f"Dropped fake Mongo database: {db_name}")


def create_fake_data_path() -> str:
    """
    Create a temporary directory to use as a location for data for a fake instance.

    :return: the data path

    """
    data_path = str(mkdtemp(prefix=f"virtool_fake_{random_alphanumeric()}_"))
    ensure_data_dir(data_path)
    return data_path


async def create_fake_user(app: App):
    """
    Create a fake user called Bob.

    :param app: the application object

    """
    await virtool.users.db.create(app["db"], USER_ID, "hello_world", True)

    await virtool.users.db.edit(app["db"], "bob", administrator=True, force_reset=False)

    logger.debug("Created fake user")


async def create_fake_subtractions(app: App):
    """
    Create fake subtractions and their associated uploads and subtraction files.

    Two subtractions are ready for use. One has ``ready`` set to ``False`` and can be used for
    testing subtraction finalization.

    :param app: the application object

    """
    async with AsyncSession(app["pg"]) as session:
        upload = Upload(name="test.fa.gz", type="subtraction", user=USER_ID)

        session.add(upload)
        await session.flush()

        upload_id = upload.id
        upload_name = upload.name

        await session.commit()

    upload = {"id": upload_id, "name": upload_name}

    await app["db"].subtraction.insert_many(
        [
            {
                "_id": "subtraction_1",
                "name": "Subtraction 1",
                "nickname": "",
                "deleted": False,
                "ready": True,
                "file": upload,
                "user": {"id": USER_ID},
            },
            {
                "_id": "subtraction_2",
                "name": "Subtraction 2",
                "nickname": "",
                "deleted": False,
                "ready": True,
                "file": upload,
                "user": {"id": USER_ID},
            },
            {
                "_id": "subtraction_unready",
                "name": "Subtraction Unready",
                "nickname": "",
                "deleted": False,
                "ready": False,
                "file": upload,
                "user": {"id": USER_ID},
            },
        ]
    )

    logger.debug("Created fake subtractions")


async def create_fake_analysis(app: App):
    """
    Create fake analyses in the database.

    Two analyses are ready for use. One has ``ready`` set to ``False`` and
    another one has ``ready`` set to ``True`` with a ``files`` field.

    :param app: the application object

    """
    sample_id = "sample_1"
    ref_id = "reference_1"
    subtractions = [
        "subtraction_1",
        "subtraction_2",
    ]

    file = await virtool.analyses.files.create_analysis_file(app["pg"], "analysis_2", "fasta", "result.fa", 123456)
    await app["db"].analyses.insert_many([
        {
            "_id": "analysis_1",
            "workflow": "pathoscope",
            "created_at": virtool.utils.timestamp(),
            "ready": False,
            "job": {
                "id": "job_1"
            },
            "index": {
                "version": 2,
                "id": "foo"
            },
            "user": {
                "id": USER_ID
            },
            "sample": {
                "id": sample_id
            },
            "reference": {
                "id": ref_id
            },
            "subtractions": subtractions
        },
        {
            "_id": "analysis_2",
            "workflow": "pathoscope",
            "created_at": virtool.utils.timestamp(),
            "ready": True,
            "job": {
                "id": "job_2"
            },
            "index": {
                "version": 2,
                "id": "foo"
            },
            "user": {
                "id": USER_ID
            },
            "sample": {
                "id": sample_id
            },
            "reference": {
                "id": ref_id
            },
            "subtractions": subtractions,
            "files": [file]
        }
    ])

    logger.debug("Created fake analyses")


async def create_integration_test_job(app: App):
    name = "integration_test_workflow"
    rights = JobRights()

    return await virtool.jobs.db.create(
        db=app["db"],
        workflow_name=name,
        job_args={
            "sample_id": "sample_1",
            "subtraction_id": "subtraction_1",
            "ref_id": "reference_1",
        },
        user_id=USER_ID,
        rights=rights,
        job_id="integration_test_job",
    )


async def create_fake_jobs(app: App):
    await create_integration_test_job(app)

