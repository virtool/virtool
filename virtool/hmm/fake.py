import json
from pathlib import Path
from shutil import copy

import aiofiles

from virtool.fake.wrapper import FakerWrapper


async def create_fake_hmms(app):
    fake: FakerWrapper = app["fake"]

    data_path = Path(app["settings"]["data_path"])
    hmms_path = data_path / "hmm"

    example_path = Path(__file__).parent.parent.parent / "example"

    copy(example_path / "hmms/profiles.hmm", hmms_path)

    async with aiofiles.open(example_path / "hmms/annotations.json", "r") as f:
        for annotation in json.loads(await f.read()):
            await app["db"].hmm.insert_one({
                **annotation,
                "_id": fake.get_mongo_id()
            })
    