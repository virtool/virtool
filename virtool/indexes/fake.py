from logging import getLogger
from shutil import copy

from sqlalchemy.ext.asyncio import AsyncEngine

from virtool.example import example_path
from virtool.fake.wrapper import FakerWrapper
from virtool.indexes.db import create, finalize
from virtool.indexes.files import create_index_file
from virtool.types import App

logger = getLogger(__name__)

INDEX_FILES = (
    "reference.fa.gz",
    "reference.1.bt2",
    "reference.2.bt2",
    "reference.3.bt2",
    "reference.4.bt2",
    "reference.rev.1.bt2",
    "reference.rev.2.bt2"
)


async def create_fake_indexes(app: App, ref_id: str, user_id: str):
    """
    Create fake indexes.

    Currently this only creates a single fake reference that includes the latest versions of the
    accompanying fake OTUs.

    TODO: Add unfinalized fake index with accompanying OTUs.

    :param app: the application object
    :param ref_id: the ID of the parent reference
    :param user_id: the ID of the creating user

    """
    db = app["db"]
    fake: FakerWrapper = app["fake"]
    settings = app["settings"]
    pg: AsyncEngine = app["pg"]

    finalized_index_id = fake.get_mongo_id()

    await create(
        db,
        ref_id,
        user_id,
        "foo",
        finalized_index_id
    )

    data_path = settings["data_path"]

    path = data_path / "references" / ref_id / finalized_index_id
    path.mkdir(parents=True)

    example_indexes = example_path / "indexes"

    for index_file in INDEX_FILES:
        copy(example_indexes / index_file, path)

        await create_index_file(
            pg,
            finalized_index_id,
            "fasta" if index_file == "reference.fa.gz" else "bowtie2",
            index_file
        )

    await finalize(
        db,
        pg,
        ref_id,
        finalized_index_id
    )

    unfinalized_index_id = fake.get_mongo_id()

    await create(
        db,
        "reference_2",
        user_id,
        "bar",
        unfinalized_index_id
    )
    logger.debug("Created fake indexes")
