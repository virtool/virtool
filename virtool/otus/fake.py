from dataclasses import dataclass
from typing import List

import virtool.otus.isolates
import virtool.otus.sequences
from virtool.fake.wrapper import FakerWrapper
from virtool.otus.db import create_otu
from virtool.types import App

SEQUENCE = """
GTATTTTTACAACAATTACCAACAACAACAAACAACAAACAACATTACAATTACTATTTACAATTACAATGGCATACACACAGACAGCTACCACATCAG
CTTTGCTGGACACTGTCCGAGGAAACAACTCCTTGGTCAATGATCTAGCAAAGCGTCGTCTTTACGACACAGCGGTTGAAGAGTTTAACGCTCGTGACC
GCAGGCCCAAGGTGAACTTTTCAAAAGTAATAAGCGAGGAGCAGACGCTTATTGCTACCCGGGCGTATCCAGAATTCCAAATTACATTTTATAACACGC
AAAATGCCGTGCATTCGCTTGCAGGTGGATTGCGATCTTTAGAACTGGAATATCTGATGATGCAAATTCCCTACGGATCATTGACTTATGACATAGGCG
GGAATTTTGCATCGCATCTGTTCAAGGGACGAGCATATGTACACTGCTGCATGCCCAACCTGGACGTTCGAGACATCATGCGGCACGAAGGCCAGAAAG
ACAGTATTGAACTATACCTTTCTAGGCTAGAGAGAGGGGGGAAAACAGTCCCCAACTTCCAAAAGGAAGCATTTGACAGATACGCAGAAATTCCTGAAG
ACGCTGTCTGTCACAATACTTTCCAGACAATGCGACATCAGCCGATGCAGCAATCAGGCAGAGTGTATGCCATTGCGCTACACAGCATATATGACATAC
CAGCCGATGAGTTCGGGGCGGCACTCTTGAGGAAAAATGTCCATACGTGCTATGCCGCTTTCCACTTCTCTGAGAACCTGCTTCTTGAAGATTCATACG
TCAATTTGGACGAAATCAACGCGTGTTTTTCGCGCGATGGAGACAAGTTGACCTTTTCTTTTGCATCAGAGAGTACTCTTAATTATTGTCATAGTTATT
CTAATATTCTTAAGTATGTGTGCAAAACTTACTTCCCGGCCTCTAATAGAGAGGTTTACATGAAGGAGTTTTTAGTCACCAGAGTTAATACCTGGTTTT
GTAAGTTTTCTAGAATAGATACTTTTCTTTTGTACAAAGGTGTGGCCCATAAAAGTGTAGATAGTGAGCAGTTTTATACTGCAATGGAAGACGCATGGC
ATTACAAAAAGACTCTTGCAATGTGCAACAGCGAGAGAATCCTCCTTGAGGATTCATCATCAGTCAATTACTGGTTTCCCAAAATGAGGGATATGGTCA
TCGTACCATTATTCGACATTTCTTTGGAGACTAGTAAGAGGACGCGCAAGGAAGTCTTAGTGTCCAAGGATTTCGTGTTTACAGTGCTTAACCACATTC
GAACATACCAGGCGAAAGCTCTTACATACGCAAATGTTTTGTCCTTTGTCGAATCGATTCGATCGAGGGTAATCATTAACGGTGTGACAGCGAGGTCCG
AATGGGATGTGGACAAATCTTTGTTACAATCCTTGTCCATGACGTTTTACCTGCATACTAAGCTTGCCGTTCTAAAGGATGACTTACTGATTAGCAAGT
TTAGTCTCGGTTCGAAAACGGTGTGCCAGCATGTGTGGGATGAGATTTCGCTGGCGTTTGGGAACGCATTTCCCTCCGTGAAAGAGAGGCTCTTGAACA
GGAAACTTATCAGAGTGGCAGGCGACGCATTAGAGATCAGGGTGCCTGATCTATATGTGACCTTCCACGACAGATTAGTGACTGAGTACAAGGCCTCTG
TGGACATGCCTGCGCTTGACATTAGGAAGAAGATGGAAGAAACGGAAGTGATGTACAATGCACTTTCAGAGTTATCGGTGTTAAGGGAGTCTGACAAAT
TCGATGTTGATGTTTTTTCCCAGATGTGCCAATCTTTGGAAGTTGACCCAATGACGGCAGCGAAGGTTATAGTCGCGGTCATGAGCAATGAGAGCGGTC
TGACTCTCACATTTGAACGACCTACTGAGGCGAATGTTGCGCTAGCTTTACAGGATCAAGAGAAGGCTTCAGAAGGTGCTTTGGTAGTTACCTCAAGAG
AAGTTGAAGAACCGTCCATGAAGGGTTCGATGGCCAGAGGAGAGTTACAATTAGCTGGTCTTGCTGGAGATCATCCGGAGTCGTCCTATTCTAAGAACG
GCCAGCAGTTGATCAGTACAGACACATGATTAAAGCACAACCCAAGCAAAAATTGGACACTTCAATCCAAACGGAGTACCCGGCTTTGCAGACGATTGT
"""


@dataclass
class FakeSequence:
    accession: str
    definition: str
    sequence: str


class FakeOTU:

    def __init__(self, app: App, ref_id: str, user_id: str, document: dict):
        self._app = app
        self._ref_id = ref_id
        self._user_id = user_id
        self._document = document

        self.otu_id = document["id"]
        self._faker: FakerWrapper = self._app["fake"]

    async def add_isolate(self, source_type, source_name, sequences: List[FakeSequence]):
        isolate_data = {
            "default": False,
            "source_type": source_type,
            "source_name": source_name
        }

        isolate_id = self._faker.get_mongo_id()

        await virtool.otus.isolates.add(
            self._app,
            self.otu_id,
            isolate_data,
            self._user_id,
            isolate_id
        )

        for sequence in sequences:
            sequence_data = {
                "accession": sequence.accession,
                "definition": sequence.definition,
                "sequence": sequence.sequence
            }

            await virtool.otus.sequences.create(
                self._app,
                self._ref_id,
                self.otu_id,
                isolate_id,
                sequence_data,
                self._user_id,
                self._faker.get_mongo_id()
            )


class FakeOTUCreator:

    def __init__(self, app: App, ref_id: str, user_id: str):
        self._app = app
        self._faker = app["fake"]
        self._ref_id = ref_id
        self._user_id = user_id

    async def create(self, name: str, abbreviation: str):
        otu_id = self._faker.get_mongo_id()

        document = await create_otu(
            self._app,
            self._ref_id,
            name,
            abbreviation,
            self._user_id,
            otu_id
        )

        return FakeOTU(self._app, self._ref_id, self._user_id, document)


async def create_fake_otus(app: App, ref_id: str, user_id: str):
    """
    Create a set of fake OTUs to populate the reference identified by ``ref_id``.

    :param app: the application object
    :param ref_id: the parent reference ID
    :param user_id: the ID of the user creating the OTUs

    """
    fake_otu_creator = FakeOTUCreator(app, ref_id, user_id)

    abtv = await fake_otu_creator.create("Abaca bunchy top virus", "ABTV")

    await abtv.add_isolate("Isolate", "A", [
        FakeSequence("foo", "ABTV sequence 1", SEQUENCE),
        FakeSequence("foo", "ABTV sequence 2", SEQUENCE),
        FakeSequence("foo", "ABTV sequence 3", SEQUENCE),
        FakeSequence("foo", "ABTV sequence 4", SEQUENCE),
        FakeSequence("foo", "ABTV sequence 5", SEQUENCE)
    ])

    aspv = await fake_otu_creator.create("Apple stem pitting virus", "ASPV")

    await aspv.add_isolate("Isolate", "Z", [FakeSequence("bar", "ASPV sequence 1", SEQUENCE)])

    tmv = await fake_otu_creator.create("Tobacco mosaic virus", "TMV")

    await tmv.add_isolate("Isolate", "1", [FakeSequence("bar", "TMV sequence 1", SEQUENCE)])
    await tmv.add_isolate("Isolate", "2", [FakeSequence("baz", "TMV sequence 1", SEQUENCE)])

    lchv_1 = await fake_otu_creator.create("Little cherry virus 1", "LChV1")

    await lchv_1.add_isolate("Isolate", "A", [FakeSequence("bar", "LChV1 sequence 1", SEQUENCE)])
    await lchv_1.add_isolate("Isolate", "B", [FakeSequence("baz", "LChV1 sequence 1", SEQUENCE)])
    await lchv_1.add_isolate("Isolate", "C", [FakeSequence("bat", "LChV1 sequence 1", SEQUENCE)])
