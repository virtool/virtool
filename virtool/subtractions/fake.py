from logging import getLogger
from pathlib import Path
from shutil import copytree
from typing import Tuple

from sqlalchemy.ext.asyncio import AsyncSession

import virtool.subtractions.db
from virtool.example import example_path
from virtool.fake.identifiers import USER_ID
from virtool.fake.wrapper import FakerWrapper
from virtool.subtractions.db import finalize
from virtool.subtractions.files import create_subtraction_files
from virtool.subtractions.utils import FILES
from virtool.types import App
from virtool.uploads.models import Upload

logger = getLogger(__name__)


async def create_fake_subtractions(app: App):
    """
    Create fake subtractions and their associated uploads and subtraction files.

    Two subtractions are ready for use. One has ``ready`` set to ``False`` and can be used for
    testing subtraction finalization.

    :param app: the application object

    """
    db = app["db"]
    fake: FakerWrapper = app["fake"]

    upload_id, upload_name = await create_fake_fasta_upload(app, USER_ID)

    await create_fake_finalized_subtraction(
        app,
        upload_id,
        upload_name,
        fake.get_mongo_id(),
        USER_ID
    )

    await virtool.subtractions.db.create(
        db,
        USER_ID,
        upload_name,
        "subtraction_2",
        "",
        upload_id,
        fake.get_mongo_id()
    )

    await virtool.subtractions.db.create(
        db,
        USER_ID,
        upload_name,
        "subtraction_unready",
        "",
        upload_id,
        fake.get_mongo_id(),
    )

    logger.debug("Created fake subtractions")


async def create_fake_fasta_upload(app: App, user_id: str) -> Tuple[int, str]:
    async with AsyncSession(app["pg"]) as session:
        upload = Upload(name="test.fa.gz", type="subtraction", user=user_id)

        session.add(upload)
        await session.flush()

        upload_id = upload.id
        upload_name = upload.name

        await session.commit()

    return upload_id, upload_name


async def create_fake_finalized_subtraction(
        app: App,
        upload_id: int,
        upload_name: str,
        subtraction_id: str,
        user_id: str
):
    db = app["db"]
    pg = app["pg"]

    document = await db.subtraction.insert_one({
        "_id": subtraction_id,
        "name": "subtraction_1",
        "nickname": "",
        "deleted": False,
        "is_host": True,
        "ready": True,
        "file": {"id": upload_id, "name": upload_name},
        "user": {"id": user_id},
    })

    subtractions_path = Path(app["settings"]["data_path"]) / "subtractions"

    subtractions_example_path = example_path / "subtractions" / "arabidopsis_thaliana"

    copytree(subtractions_example_path, subtractions_path / subtraction_id, dirs_exist_ok=True)

    await create_subtraction_files(
        pg,
        document["_id"],
        FILES,
        subtractions_path / subtraction_id
    )

    await finalize(
        db,
        pg,
        subtraction_id,
        {
            "a": 0.25,
            "t": 0.25,
            "g": 0.25,
            "c": 0.25
        }
    )
