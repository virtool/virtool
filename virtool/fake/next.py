"""
Easily create fake data.

"""
import asyncio
import gzip
from pathlib import Path
from typing import Type, AsyncGenerator

import aiofiles
from faker import Faker
from faker.providers import BaseProvider, python, color, lorem, file
from sqlalchemy import update
from sqlalchemy.ext.asyncio import AsyncEngine, AsyncSession
from virtool_core.models.group import Group
from virtool_core.models.hmm import HMM
from virtool_core.models.job import Job, JobState
from virtool_core.models.label import Label
from virtool_core.models.roles import AdministratorRole
from virtool_core.models.task import Task
from virtool_core.models.upload import Upload
from virtool_core.models.user import User

from virtool.data.layer import DataLayer
from virtool.example import example_path
from virtool.groups.oas import UpdateGroupRequest, PermissionsUpdate
from virtool.groups.pg import SQLGroup
from virtool.indexes.tasks import EnsureIndexFilesTask
from virtool.jobs.utils import WORKFLOW_NAMES
from virtool.ml.models import MLModel
from virtool.mongo.core import Mongo
from virtool.references.tasks import CleanReferencesTask, CloneReferenceTask
from virtool.releases import ReleaseManifestItem
from virtool.subtractions.tasks import AddSubtractionFilesTask
from virtool.tasks.task import BaseTask
from virtool.uploads.models import UploadType
from virtool.uploads.utils import CHUNK_SIZE
from virtool.users.oas import UpdateUserRequest


async def fake_file_chunker(path: Path) -> AsyncGenerator[bytearray, None]:
    """
    Read a chunk of size `CHUNK_SIZE` from a file.

    This is used to hijack the upload process and create fake uploads.

    :param path: the path to the file
    :return: an async generator that yields chunks of the file
    """
    async with aiofiles.open(path, "rb") as f:
        yield await f.read(CHUNK_SIZE)


async def gzip_file_chunker(path: Path) -> AsyncGenerator[bytearray, None]:
    """
    Decompress and Read a chunk of size `CHUNK_SIZE` from a file.

    This is used to hijack the upload process and create fake uploads. It will
    decompress the file at ``path`` before reading it.

    :param path: the path to the gzip file
    :return: an async generator that yields chunks of the file

    """
    _q = asyncio.Queue()

    def reader(gzip_path: Path, q: asyncio.Queue):
        with gzip.open(gzip_path, "rb") as f:
            q.put_nowait(f.read(CHUNK_SIZE))

    task = asyncio.Task(asyncio.to_thread(reader, path, _q))

    while not _q.empty():
        yield await _q.get()

    await task


class VirtoolProvider(BaseProvider):
    def mongo_id(self):
        return self.random_letters(8)

    def pg_id(self) -> int:
        return self.random_int()

    def workflow(self) -> str:
        return self.random_choices(
            [name.replace("job_", "") for name in WORKFLOW_NAMES], 1
        )[0]


class DataFaker:
    def __init__(self, layer: DataLayer, mongo: Mongo, pg: AsyncEngine):
        self.layer = layer
        self.mongo = mongo
        self.pg = pg

        self.faker = Faker()
        self.faker.seed_instance(0)
        self.faker.add_provider(VirtoolProvider)
        self.faker.add_provider(color)
        self.faker.add_provider(lorem)
        self.faker.add_provider(python)
        self.faker.add_provider(file)

        self.groups = GroupsFakerPiece(self)
        self.hmm = HMMFakerPiece(self)
        self.jobs = JobsFakerPiece(self)
        self.labels = LabelsFakerPiece(self)
        self.ml = MLFakerPiece(self)
        self.tasks = TasksFakerPiece(self)
        self.users = UsersFakerPiece(self)
        self.uploads = UploadsFakerPiece(self)

        self.mongo = mongo


class DataFakerPiece:
    def __init__(self, data_faker: DataFaker):
        self._faker = data_faker.faker
        self._layer: DataLayer = data_faker.layer
        self._mongo = data_faker.mongo
        self._pg = data_faker.pg

        self.history = []


class JobsFakerPiece(DataFakerPiece):
    model = Job

    async def create(
        self,
        user: User,
        archived: bool = False,
        state: JobState | None = None,
        workflow: str | None = None,
    ) -> Job:
        """
        Create a fake job.

        A completely valid job will be created which follows the rules:

        * The job will be in an archive-able state if ``archived`` is ``True``.
        * The job status updates will be in a valid order.

        :param user: the user that created the job
        :param archived: whether the job should be archived
        :param state: the state the most recent status update should have
        :param workflow: the workflow the job is running
        :return:
        """
        if archived and state in [
            JobState.WAITING,
            JobState.PREPARING,
            JobState.RUNNING,
        ]:
            raise ValueError(
                "Cannot create an archived job in the waiting, preparing, or running states"
            )

        job = await self._layer.jobs.create(
            workflow or self._faker.workflow(),
            self._faker.pydict(nb_elements=6, value_types=[str, int, float]),
            user.id,
            0,
        )

        if state:
            target_state = state
        elif archived:
            target_state = self._faker.random_element(
                [
                    JobState.CANCELLED,
                    JobState.COMPLETE,
                    JobState.ERROR,
                    JobState.TERMINATED,
                    JobState.TIMEOUT,
                ]
            )
        else:
            target_state = self._faker.random_element(
                [
                    JobState.CANCELLED,
                    JobState.COMPLETE,
                    JobState.ERROR,
                    JobState.PREPARING,
                    JobState.RUNNING,
                    JobState.TERMINATED,
                    JobState.TIMEOUT,
                    JobState.WAITING,
                ]
            )

        previous_status = job.status[0]
        progress = 0

        while progress <= 50:
            if previous_status.state == target_state:
                break

            if progress == 50:
                next_state = target_state
            elif previous_status.state == JobState.WAITING:
                next_state = JobState.PREPARING
            elif previous_status.state == JobState.PREPARING:
                next_state = JobState.RUNNING
            else:
                next_state = self._faker.random_element(
                    [JobState.RUNNING, target_state]
                )

            progress += 10
            step_name = self._faker.pystr()

            previous_status = await self._layer.jobs.push_status(
                job.id,
                state=next_state,
                stage=step_name,
                step_name=step_name,
                step_description=self._faker.pystr(),
                error=None,
                progress=100 if next_state == JobState.COMPLETE else progress,
            )

        if archived:
            return await self._layer.jobs.archive(job.id)

        return await self._layer.jobs.get(job.id)


class GroupsFakerPiece(DataFakerPiece):
    model = Group

    async def create(
        self, permissions: PermissionsUpdate | None = None, legacy_id: str | None = None
    ) -> Group:
        """
        :param permissions: Permissions for the group. If not provided, the group will
            have no permissions.
        :param legacy_id: An optional legacy ID for the group.
        :return: a group
        """
        name = "contains spaces"

        while " " in name:
            name = self._faker.profile()["job"].lower() + "s"

        group = await self._layer.groups.create(name)

        if legacy_id:
            async with AsyncSession(self._pg) as session:
                await session.execute(
                    update(SQLGroup)
                    .where(SQLGroup.id == group.id)
                    .values(legacy_id=legacy_id)
                )
                await session.commit()

            group = await self._layer.groups.get(group.id)

        if permissions:
            group = await self._layer.groups.update(
                group.id, UpdateGroupRequest(permissions=permissions)
            )

        return group


class HMMFakerPiece(DataFakerPiece):
    model = HMM

    async def create(self, mongo: Mongo) -> HMM:
        """
        Create a new fake hmm.

        :param mongo: the mongo DB connection

        :return: a new fake hmm
        """
        hmm_id = "".join(self._faker.mongo_id())
        faker = self._faker

        document = await mongo.hmm.insert_one(
            {
                "entries": [
                    {
                        "accession": faker.pystr(),
                        "gi": faker.pystr(),
                        "name": faker.pystr(),
                        "organism": faker.pystr(),
                    }
                ],
                "genera": {faker.pystr(): faker.pyint()},
                "length": faker.pyint(),
                "mean_entropy": faker.pyfloat(),
                "total_entropy": faker.pyfloat(),
                "_id": hmm_id,
                "cluster": faker.pyint(),
                "count": faker.pyint(),
                "families": {faker.pystr(): faker.pyint()},
                "names": [faker.pystr()],
            }
        )

        return HMM(**document)


class LabelsFakerPiece(DataFakerPiece):
    model = Label

    async def create(self) -> Label:
        """
        Create a new fake label.

        :return: a new fake label
        """
        return await self._layer.labels.create(
            self._faker.word().capitalize(),
            self._faker.hex_color(),
            self._faker.sentence(),
        )


class MLFakerPiece(DataFakerPiece):
    model = MLModel

    async def populate(self):
        """Populate the ML model collection with fake data."""
        return await self._layer.ml.load(
            {
                "ml-plant-viruses": [
                    self.create_release_manifest_item() for _ in range(3)
                ],
                "ml-insect-viruses": [
                    self.create_release_manifest_item() for _ in range(3)
                ],
            },
        )

    def create_release_manifest_item(self) -> ReleaseManifestItem:
        """
        Create a fake release manifest item like you would receive from the
        www.virtool.ca releases endpoints

        """
        return ReleaseManifestItem(
            id=self._faker.pyint(),
            body=self._faker.paragraph(),
            content_type=self._faker.pystr(),
            download_url=self._faker.url(),
            filename=self._faker.pystr(),
            html_url=self._faker.url(),
            name=self._faker.pystr(),
            prerelease=self._faker.pybool(),
            published_at=self._faker.date_time(),
            size=self._faker.pyint(),
        )


class TasksFakerPiece(DataFakerPiece):
    model = Task

    async def create(self) -> Task:
        """
        Create a new fake random task.

        :return: a new fake task
        """
        return await self._layer.tasks.create(
            self._faker.random_element(
                [
                    EnsureIndexFilesTask,
                    AddSubtractionFilesTask,
                    CloneReferenceTask,
                    CleanReferencesTask,
                ]
            ),
            {},
        )

    async def create_with_class(self, cls: Type[BaseTask], context: dict) -> Task:
        """
        Create a fake task.

        :param cls: the type of task
        :param context: the context required for the task

        :return: a new fake task
        """
        return await self._layer.tasks.create(cls, context)


class UploadsFakerPiece(DataFakerPiece):
    model = Upload

    async def create(
        self,
        user: User,
        upload_type: UploadType = UploadType.reads,
        name: str = "test.fq.gz",
        reserved: bool = False,
    ) -> Upload:
        """
        Create a fake upload.

        A completely valid user will be created.

        :param user: the user creating the upload
        :param upload_type: the type of upload
        :param name: the name of the upload
        :param reserved: the reservation status of the upload

        :return: a new fake upload
        """
        if upload_type not in UploadType.to_list():
            upload_type = "reads"

        fake_file_path = example_path / "reads/single.fq.gz"

        upload = await self._layer.uploads.create(
            fake_file_chunker(fake_file_path), name, upload_type, user.id
        )

        if reserved:
            await self._layer.uploads.reserve(upload.id)

        return await self._layer.uploads.get(upload.id)


class UsersFakerPiece(DataFakerPiece):
    model = User

    async def create(
        self,
        handle: str | None = None,
        groups: list[Group] | None = None,
        password: str | None = None,
        primary_group: Group | None = None,
        administrator_role: AdministratorRole | None = None,
    ) -> User:
        """
        Create a fake user.

        A completely valid user will be created.

        :param handle: the users handle
        :param groups: the groups the user belongs to
        :param password: the users password
        :param primary_group: the users primary group
        :param administrator_role: the users administrator role

        :return: a new fake user
        """
        user = await self._layer.users.create(
            handle or self._faker.profile()["username"],
            password or self._faker.password(),
        )

        if administrator_role:
            user = await self._layer.users.set_administrator_role(
                user.id, administrator_role
            )

        if groups and primary_group:
            return await self._layer.users.update(
                user.id,
                UpdateUserRequest(
                    groups=list({group.id for group in groups} | {primary_group.id})
                ),
            )

        if groups:
            return await self._layer.users.update(
                user.id,
                UpdateUserRequest(groups=[group.id for group in groups]),
            )

        if primary_group:
            return await self._layer.users.update(
                user.id,
                UpdateUserRequest(
                    groups=[primary_group.id], primary_group=primary_group.id
                ),
            )

        return user
