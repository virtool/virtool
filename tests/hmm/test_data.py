from virtool.fake.next import DataFaker


async def test_get_status(config, data_layer, fake: DataFaker, mongo, snapshot, static_time):
    """Test that function works when the HMM data are being updated and when they are not.

    """
    user = await fake.users.create()

    await mongo.status.insert_one(
        {
            "_id": "hmm",
            "updating": False,
            "updates": [{"id": 231}],
            "installed": {
                "body": "- remove some annotations that didn't have corresponding profiles",
                "created_at": static_time.datetime,
                "filename": "vthmm.tar.gz",
                "html_url": "https://github.com/virtool/virtool-hmm/releases/tag/v0.2.1",
                "id": 8472569,
                "name": "v0.2.1",
                "newer": True,
                "published_at": "2017-11-10T19:12:43Z",
                "ready": True,
                "size": 85904451,
                "user": {"id": user.id},
            },
            "release": {
                "body": "- remove some annotations that didn't have corresponding profiles",
                "content_type": "application/gzip",
                "download_url": "https://github.com/virtool/virtool-hmm/releases/download/v0.2.1/vthmm.tar.gz",
                "etag": 'W/"7bd9cdef79c82ab4d7e5cfff394cf81eaddc6f681b8202f2a7bdc65cbcc4aaea"',
                "filename": "vthmm.tar.gz",
                "html_url": "https://github.com/virtool/virtool-hmm/releases/tag/v0.2.1",
                "id": 1230982,
                "name": "v0.2.1",
                "newer": False,
                "published_at": static_time.datetime,
                "retrieved_at": static_time.datetime,
                "size": 85904451,
            },
            "errors": [],
        },
    )

    assert await data_layer.hmms.get_status() == snapshot
