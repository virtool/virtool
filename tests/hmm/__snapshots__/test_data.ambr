# serializer version: 1
# name: test_get_status
  dict({
    'errors': list([
    ]),
    'installed': dict({
      'body': "- remove some annotations that didn't have corresponding profiles",
      'created_at': datetime.datetime(2015, 10, 6, 20, 0),
      'filename': 'vthmm.tar.gz',
      'html_url': 'https://github.com/virtool/virtool-hmm/releases/tag/v0.2.1',
      'id': 8472569,
      'name': 'v0.2.1',
      'newer': True,
      'published_at': datetime.datetime(2017, 11, 10, 19, 12, 43, tzinfo=datetime.timezone.utc),
      'ready': True,
      'size': 85904451,
      'user': dict({
        'handle': 'leeashley',
        'id': 'bf1b993c',
      }),
    }),
    'release': dict({
      'body': "- remove some annotations that didn't have corresponding profiles",
      'content_type': 'application/gzip',
      'download_url': 'https://github.com/virtool/virtool-hmm/releases/download/v0.2.1/vthmm.tar.gz',
      'etag': 'W/"7bd9cdef79c82ab4d7e5cfff394cf81eaddc6f681b8202f2a7bdc65cbcc4aaea"',
      'filename': 'vthmm.tar.gz',
      'html_url': 'https://github.com/virtool/virtool-hmm/releases/tag/v0.2.1',
      'id': 1230982,
      'name': 'v0.2.1',
      'newer': False,
      'published_at': datetime.datetime(2015, 10, 6, 20, 0),
      'retrieved_at': datetime.datetime(2015, 10, 6, 20, 0),
      'size': 85904451,
    }),
    'task': None,
    'updating': False,
  })
# ---
